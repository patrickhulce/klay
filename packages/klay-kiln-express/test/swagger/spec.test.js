const buildSpecification = require('../../dist/swagger/spec').buildSpecification
const CRUD_ROUTES = require('../../dist/extensions/router').CRUD_ROUTES
const utils = require('../utils')

describe('lib/swagger/spec.ts', () => {
  let kiln

  beforeEach(() => {
    kiln = utils.state().kiln
  })

  describe('#buildSpec', () => {
    it('should build a specification', () => {
      const routerOpts = {
        routes: CRUD_ROUTES
      }

      const router = kiln.build('user', 'express-router', routerOpts)
      const spec = buildSpecification(kiln, router)
      expect(spec).toMatchSnapshot()
    })
  })
})
