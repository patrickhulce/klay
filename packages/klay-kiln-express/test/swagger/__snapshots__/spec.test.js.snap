// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/swagger/spec.ts #buildSpec should build a specification 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "BooleanFilters": Object {
      "properties": Object {
        "$eq": Object {
          "type": "boolean",
        },
        "$in": Object {
          "items": Object {
            "type": "boolean",
          },
          "type": "array",
        },
        "$ne": Object {
          "type": "boolean",
        },
        "$nin": Object {
          "items": Object {
            "type": "boolean",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "CreateUserPayload": Object {
      "properties": Object {
        "age": Object {
          "format": "int64",
          "type": "integer",
        },
        "email": Object {
          "maxLength": 250,
          "type": "string",
        },
        "firstName": Object {
          "maxLength": 100,
          "type": "string",
        },
        "isAdmin": Object {
          "type": "boolean",
        },
        "lastName": Object {
          "maxLength": 100,
          "type": "string",
        },
        "metadata": Object {
          "$ref": "#/definitions/UserMetadata",
        },
        "password": Object {
          "maxLength": 32,
          "type": "string",
        },
        "tracking": Object {
          "$ref": "#/definitions/CreateUserPayloadTracking",
        },
      },
      "required": Array [
        "age",
        "isAdmin",
        "email",
        "password",
        "firstName",
        "lastName",
        "tracking",
      ],
      "type": "object",
    },
    "CreateUserPayloadTracking": Object {
      "properties": Object {
        "lastVisit": Object {
          "format": "date-time",
          "type": "string",
        },
      },
      "type": "object",
    },
    "CreateUsersPayload": Object {
      "items": Object {
        "$ref": "#/definitions/CreateUserPayload",
      },
      "type": "array",
    },
    "DateFilters": Object {
      "properties": Object {
        "$eq": Object {
          "format": "date-time",
          "type": "string",
        },
        "$gt": Object {
          "format": "date-time",
          "type": "string",
        },
        "$gte": Object {
          "format": "date-time",
          "type": "string",
        },
        "$in": Object {
          "items": Object {
            "format": "date-time",
            "type": "string",
          },
          "type": "array",
        },
        "$lt": Object {
          "format": "date-time",
          "type": "string",
        },
        "$lte": Object {
          "format": "date-time",
          "type": "string",
        },
        "$ne": Object {
          "format": "date-time",
          "type": "string",
        },
        "$nin": Object {
          "items": Object {
            "format": "date-time",
            "type": "string",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "DestroyUsersPayload": Object {
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
    "ListUsersPayload": Object {
      "properties": Object {
        "age": Object {
          "$ref": "#/definitions/NumberFilters",
        },
        "createdAt": Object {
          "$ref": "#/definitions/DateFilters",
        },
        "email": Object {
          "$ref": "#/definitions/StringFilters",
        },
        "fields": Object {
          "type": "string",
        },
        "firstName": Object {
          "$ref": "#/definitions/StringFilters",
        },
        "id": Object {
          "$ref": "#/definitions/StringFilters",
        },
        "isAdmin": Object {
          "$ref": "#/definitions/BooleanFilters",
        },
        "lastName": Object {
          "$ref": "#/definitions/StringFilters",
        },
        "limit": Object {
          "default": 10,
          "format": "int32",
          "maximum": 1000,
          "type": "integer",
        },
        "offset": Object {
          "default": 0,
          "format": "int64",
          "minimum": 0,
          "type": "integer",
        },
        "order": Object {
          "type": "string",
        },
        "password": Object {
          "$ref": "#/definitions/StringFilters",
        },
        "updatedAt": Object {
          "$ref": "#/definitions/DateFilters",
        },
      },
      "type": "object",
    },
    "ListUsersResponse": Object {
      "properties": Object {
        "data": Object {
          "$ref": "#/definitions/UsersList",
        },
        "total": Object {
          "format": "int64",
          "type": "integer",
        },
      },
      "required": Array [
        "data",
        "total",
      ],
      "type": "object",
    },
    "NumberFilters": Object {
      "properties": Object {
        "$eq": Object {
          "format": "int64",
          "type": "integer",
        },
        "$in": Object {
          "items": Object {
            "format": "int64",
            "type": "integer",
          },
          "type": "array",
        },
        "$ne": Object {
          "format": "int64",
          "type": "integer",
        },
        "$nin": Object {
          "items": Object {
            "format": "int64",
            "type": "integer",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "StringFilters": Object {
      "properties": Object {
        "$eq": Object {
          "type": "string",
        },
        "$in": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "$match": Object {
          "type": "string",
        },
        "$ne": Object {
          "type": "string",
        },
        "$nin": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "UpdateUserPayload": Object {
      "properties": Object {
        "age": Object {
          "format": "int64",
          "type": "integer",
        },
        "createdAt": Object {
          "format": "date-time",
          "type": "string",
        },
        "email": Object {
          "maxLength": 250,
          "type": "string",
        },
        "firstName": Object {
          "maxLength": 100,
          "type": "string",
        },
        "id": Object {
          "type": "string",
        },
        "isAdmin": Object {
          "type": "boolean",
        },
        "lastName": Object {
          "maxLength": 100,
          "type": "string",
        },
        "metadata": Object {
          "$ref": "#/definitions/UserMetadata",
        },
        "password": Object {
          "maxLength": 32,
          "type": "string",
        },
        "tracking": Object {
          "$ref": "#/definitions/UpdateUserPayloadTracking",
        },
        "updatedAt": Object {
          "format": "date-time",
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "age",
        "isAdmin",
        "email",
        "password",
        "firstName",
        "lastName",
        "tracking",
      ],
      "type": "object",
    },
    "UpdateUserPayloadTracking": Object {
      "properties": Object {
        "lastVisit": Object {
          "format": "date-time",
          "type": "string",
        },
        "trackingId": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "UpdateUsersPayload": Object {
      "items": Object {
        "$ref": "#/definitions/UpdateUserPayload",
      },
      "type": "array",
    },
    "User": Object {
      "properties": Object {
        "age": Object {
          "format": "int64",
          "type": "integer",
        },
        "createdAt": Object {
          "format": "date-time",
          "type": "string",
        },
        "email": Object {
          "maxLength": 250,
          "type": "string",
        },
        "firstName": Object {
          "maxLength": 100,
          "type": "string",
        },
        "id": Object {
          "type": "string",
        },
        "isAdmin": Object {
          "type": "boolean",
        },
        "lastName": Object {
          "maxLength": 100,
          "type": "string",
        },
        "metadata": Object {
          "$ref": "#/definitions/UserMetadata",
        },
        "password": Object {
          "maxLength": 32,
          "type": "string",
        },
        "tracking": Object {
          "$ref": "#/definitions/UserTracking",
        },
        "updatedAt": Object {
          "format": "date-time",
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "age",
        "isAdmin",
        "email",
        "password",
        "firstName",
        "lastName",
        "tracking",
        "createdAt",
        "updatedAt",
      ],
      "type": "object",
    },
    "UserMetadata": Object {
      "properties": Object {},
      "type": "object",
    },
    "UserTracking": Object {
      "properties": Object {
        "lastVisit": Object {
          "format": "date-time",
          "type": "string",
        },
        "trackingId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "trackingId",
      ],
      "type": "object",
    },
    "UsersList": Object {
      "items": Object {
        "$ref": "#/definitions/User",
      },
      "type": "array",
    },
  },
  "host": "localhost",
  "info": Object {
    "title": "Title",
    "version": "Version",
  },
  "paths": Object {
    "/": Object {
      "delete": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "204": Object {
            "description": "Destroy User was successful",
            "schema": undefined,
          },
        },
        "summary": "Destroy User",
        "tags": Array [
          "user",
        ],
      },
      "get": Object {
        "parameters": Array [
          Object {
            "default": 10,
            "format": "int32",
            "in": "query",
            "maximum": 1000,
            "name": "limit",
            "required": false,
            "type": "integer",
          },
          Object {
            "default": 0,
            "format": "int64",
            "in": "query",
            "minimum": 0,
            "name": "offset",
            "required": false,
            "type": "integer",
          },
          Object {
            "in": "query",
            "name": "order",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "fields",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "id[$eq]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "id[$ne]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "id[$match]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "id[$in][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "id[$nin][]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "int64",
            "in": "query",
            "name": "age[$eq]",
            "required": false,
            "type": "integer",
          },
          Object {
            "format": "int64",
            "in": "query",
            "name": "age[$ne]",
            "required": false,
            "type": "integer",
          },
          Object {
            "format": "int64",
            "in": "query",
            "name": "age[$in][]",
            "required": false,
            "type": "integer",
          },
          Object {
            "format": "int64",
            "in": "query",
            "name": "age[$nin][]",
            "required": false,
            "type": "integer",
          },
          Object {
            "in": "query",
            "name": "isAdmin[$eq]",
            "required": false,
            "type": "boolean",
          },
          Object {
            "in": "query",
            "name": "isAdmin[$ne]",
            "required": false,
            "type": "boolean",
          },
          Object {
            "in": "query",
            "name": "isAdmin[$in][]",
            "required": false,
            "type": "boolean",
          },
          Object {
            "in": "query",
            "name": "isAdmin[$nin][]",
            "required": false,
            "type": "boolean",
          },
          Object {
            "in": "query",
            "name": "email[$eq]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "email[$ne]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "email[$match]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "email[$in][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "email[$nin][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "password[$eq]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "password[$ne]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "password[$match]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "password[$in][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "password[$nin][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "firstName[$eq]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "firstName[$ne]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "firstName[$match]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "firstName[$in][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "firstName[$nin][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "lastName[$eq]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "lastName[$ne]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "lastName[$match]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "lastName[$in][]",
            "required": false,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "lastName[$nin][]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$eq]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$ne]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$lt]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$gt]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$gte]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$lte]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$in][]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "createdAt[$nin][]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$eq]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$ne]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$lt]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$gt]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$gte]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$lte]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$in][]",
            "required": false,
            "type": "string",
          },
          Object {
            "format": "date-time",
            "in": "query",
            "name": "updatedAt[$nin][]",
            "required": false,
            "type": "string",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "List Users was successful",
            "schema": Object {
              "$ref": "#/definitions/ListUsersResponse",
            },
          },
        },
        "summary": "List Users",
        "tags": Array [
          "user",
        ],
      },
      "post": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/CreateUserPayload",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Create User was successful",
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": "Create User",
        "tags": Array [
          "user",
        ],
      },
      "put": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/UpdateUserPayload",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Update User was successful",
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": "Update User",
        "tags": Array [
          "user",
        ],
      },
    },
    "/:id": Object {
      "delete": Object {
        "parameters": Array [
          Object {
            "in": "params",
            "name": "id",
            "required": true,
            "type": "string",
          },
        ],
        "responses": Object {
          "204": Object {
            "description": "Destroy User was successful",
            "schema": undefined,
          },
        },
        "summary": "Destroy User",
        "tags": Array [
          "user",
        ],
      },
      "get": Object {
        "parameters": Array [
          Object {
            "in": "params",
            "name": "id",
            "required": true,
            "type": "string",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Read User was successful",
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": "Read User",
        "tags": Array [
          "user",
        ],
      },
      "put": Object {
        "parameters": Array [
          Object {
            "in": "params",
            "name": "id",
            "required": true,
            "type": "string",
          },
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/UpdateUserPayload",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Update User was successful",
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": "Update User",
        "tags": Array [
          "user",
        ],
      },
    },
    "/bulk": Object {
      "delete": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/DestroyUsersPayload",
            },
          },
        ],
        "responses": Object {
          "204": Object {
            "description": "Destroy Users was successful",
            "schema": undefined,
          },
        },
        "summary": "Destroy Users",
        "tags": Array [
          "user",
        ],
      },
      "post": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/CreateUsersPayload",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Create Users was successful",
            "schema": Object {
              "$ref": "#/definitions/UsersList",
            },
          },
        },
        "summary": "Create Users",
        "tags": Array [
          "user",
        ],
      },
      "put": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/UpdateUsersPayload",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Update Users was successful",
            "schema": Object {
              "$ref": "#/definitions/UsersList",
            },
          },
        },
        "summary": "Update Users",
        "tags": Array [
          "user",
        ],
      },
    },
    "/search": Object {
      "post": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": false,
            "schema": Object {
              "$ref": "#/definitions/ListUsersPayload",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "List Users was successful",
            "schema": Object {
              "$ref": "#/definitions/ListUsersResponse",
            },
          },
        },
        "summary": "List Users",
        "tags": Array [
          "user",
        ],
      },
    },
  },
  "produces": Array [
    "application/json",
  ],
  "schemes": Array [
    "http",
  ],
  "swagger": "2.0",
}
`;
